### PR TITLE
Add support for negative values

### PR DESCRIPTION
**Pull request by gsmecher**
This is an updated version of #153, and should clear up https://github.com/tenox7/ttyplot/issues/53 and https://github.com/tenox7/ttyplot/issues/22.

This is a quick and possibly rough implementation - comments welcome. The rebase looks clean enough, but I admit to rushing through it.

### CODE REVIEW

File: ttyplot.c [hunk 1]
```diff
@@ -203,27 +201,43 @@ static void draw_line(int x, int ph, int l1, int l2, cchar_t *c1, cchar_t *c2,
...
+            l1 = 0;
+        else
+            l1 = lrint((v1[i] - min) / (max - min) * ph);
+
+        if (! v2 || isnan(v2[i]))
+            l2 = 0;
+        else if (v2[i] > hardmax)
+            l2 = ph;
+        else if (v2[i] < hardmin)
+            l2 = 0;
```
Comment (edgar-bonet) [inline ttyplot.c:228]:
```suggestion
        else if (v2[i] < hardmin)
            l2 = 1;
```
For symmetry with the case `v1[i] < hardmin`.

File: ttyplot.c [hunk 2]
```diff
@@ -233,15 +247,19 @@ static void show_all_centered(const char *message) {
     mvaddnstr(y, x, message, width);
 }
 
-static int window_big_enough_to_draw(void) {
-    return (width >= 68) && (height >= 5);
+static int window_size_suitable(void) {
+    return (width >= 68) && (height >= 5) && (width < NSAMP);
 }
```
Comment (edgar-bonet) [inline ttyplot.c:252]: If the window size is not suitable, then an error message is displayed by `show_window_size_error()`: “Window too small...”. Now that “not suitable” can also mean “too big”, the error message should be adjusted.

File: ttyplot.c [hunk 3]
```diff
@@ -203,27 +201,43 @@ static void draw_line(int x, int ph, int l1, int l2, cchar_t *c1, cchar_t *c2,
...
+
+    for (x = first_col; x < first_col + pw; x++, i = (i + 1) % pw) {
+        /* suppress drawing uninitialized entries */
+        if (! v1 || isnan(v1[i]))
+            continue;
+
+        if (v1[i] > hardmax)
+            l1 = ph;
+        else if (v1[i] < hardmin)
+            l1 = 0;
```
Comment (edgar-bonet) [inline ttyplot.c:219]:
```suggestion
        else if (v1[i] < hardmin)
            l1 = 1;
```
Ttyplot normally displays an “error symbol” when a value is less than `hardmin`. In order for this symbol to be visible, `l1` should be 1 here.

**Review by edgar-bonet — CHANGES_REQUESTED** [review]
Thanks! This looks pretty nice to me. You even fixed a couple of errors in passing... Just a couple of minor things:

File: ttyplot.c [hunk 4]
```diff
@@ -333,16 +325,18 @@ static void signal_handler(int signum) {
 }
 
 static void redraw_screen(const char *errstr) {
-    if (window_size_suitable()) {
+    if (width > NSAMP) {
```
Comment (edgar-bonet) [inline ttyplot.c:328]: This should be `plotwidth`:

* `width` is the total width of the terminal window
* `plotwidth`, which is set to `width - 4` by `paint_plot()`, is the width of the area used for plotting the data.

Within `plot_values()`, the data arrays are accessed through an index that wraps modulo `plotwidth`.

**Review by edgar-bonet — COMMENTED** [review]
There are a couple things I should have noticed in my previous review, but they only strike me now... First is the confusion between `width` and `plotwidth` (see below). The second is that I do not understand the need for the check for `plotwidth > NSAMP`, as it seems redundant with this check near the top of `paint_plot()`:

```c
if (plotwidth >= (int)((sizeof(values1) / sizeof(double)) - 1))
    exit(0);
```

I did a test: starting from the head of this PR (163450c), I made `NSAMP` much smaller and removed the test `if (width > NSAMP)`. I compiled with `-fsanitize=address`, then ran `stresstest | ttyplot`. I enlarged the window while ttyplot was running and, when it got too large, ttyplot just exited (without properly resetting the terminal :unamused:). No memory errors. Repeated a few times, with repeatable results.

In the commit log of d1b0fa6, you wrote “Fix segfault with very wide windows”. Did you actually witness such segfault? If not, I suggest you drop this test for now.

There is still something that doesn't look quite right about the size checks, but it is the same here and on master, so I think it could be left for a subsequent pull request. While redrawing the screen, the normal sequence of events is:

1. `redraw_screen()` checks for the window not being too small (and also too large in this PR)
2. `paint_plot()` calls `getmaxyx()` to update our idea of the window size
3. `paint_plot()` checks for the window not being too large

Items 1 and 2 should be reversed in order for both checks to use the updated window size, as this is the one that will be used when actually plotting. There is also a race condition here if the window gets too small between 1 and 2, although the window of vulnerability seems tiny.

File: ttyplot.c (hunk 4 above)
Comment (gsmecher) [reply ttyplot.c:328]: Hm, you are correct - this part of the PR looks misguided.

I do remember triggering a segfault, but in hindsight it's possible I triggered the exit(0) path and mistook it for something worse without following up.

This feels like a distraction - unless you have a preference, I think I'll back the overrange check out of the negative-value PR unless I can more carefully trigger a bug.

**Comment by hartwork** [comment]
@gsmecher just a quick note that with the next push your CI here will be red due to meantime changes at https://apt.llvm.org/ *unless* you rebase onto latest `master` where pull request #170 with a fix has been merged. Thanks!

**Comment by gsmecher** [comment]
> @gsmecher just a quick note that with the next push your CI here will be red due to meantime changes at https://apt.llvm.org/ _unless_ you rebase onto latest `master` where pull request #170 with a fix has been merged. Thanks!

Thanks - I have rebased and force-pushed, removing the bogus array size check at the same time.

**Comment by edgar-bonet** [comment]
@gsmecher: There is something wrong with your last force-push. The previous one (77fb446) was, as expected, based on master. The last one (3d83c57) is based on master~ and carries copies of the two commits of #170, with different Committer name:

```text
* 3d83c57 (pull/169/head) Add support for negative values.
* 622f3a9 stresstest: add -n flag to remove positive bias (emit negative values)
* 6d24924 Don't use FLT_MIN - 0 is the correct value here.
* 1c579ef linux_and_macos.yml: Drop compilation with Clang 17
* bb6a97d linux_and_macos.yml: Get installation of Clang 18 on Linux back in sy>
* 5b64245 Move global variables that are only locally used into local scope.
| * 77fb446 Add support for negative values.
| * 0f1ae18 stresstest: add -n flag to remove positive bias (emit negative valu>
| * 216be4a Don't use FLT_MIN - 0 is the correct value here.
| * 0eb1356 Move global variables that are only locally used into local scope.
| * f98f2f2 (master) Merge pull request #170 from hartwork/fix-linux-ci
|/| 
| * 321ebd5 linux_and_macos.yml: Drop compilation with Clang 17
| * aef9d29 linux_and_macos.yml: Get installation of Clang 18 on Linux back in >
|/  
* 374bca3 readme update
```

Also, I had to unresolve a couple of review comments.

**Comment by gsmecher** [comment]
My apologies. I am trying to keep this as a "side distraction" - which isn't fair if it pushes work onto you.

I have rebased onto f98f, and can see the correct commit history since then. I'll force-push that now, then look into the resolved/unresolved comments. Thanks for bearing with me.

**Review by edgar-bonet — APPROVED** [review]

**Review by tenox7 — APPROVED** [review]
Thank you!!! LGTM

**Comment by tenox7** [comment]
Do you want me to merge?

**Review by tenox7 — APPROVED** [review]
LGTM

**Comment by tenox7** [comment]
Thank you!

**Pull request merged**